### PR TITLE
planner: fix ```can't find column``` for count_star_rewriter

### LINKED ISSUE #39506: ```Can't find column``` when count star rewriter is enable
## Bug Report

```
MySQL [test]> explain select count(*) from github_events_1;
ERROR 1105 (HY000): Can't find column test.github_events_1.test_count in schema Column: [Column#8] Unique key: []
```

### 1. Minimal reproduce step (Required)

1. create table with a not null and not key column
```
CREATE TABLE `github_events_1` (
  `id` bigint(20) DEFAULT NULL,
  `test_count` tinyint(1) NOT NULL DEFAULT '1',
  KEY `index_github_events_on_id` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
```
2. add a tiflash replica
```
alter table github_events_1 set tiflash replica 1;
```
3. remove tiflash read engine
```
set tidb_isolation_read_engines = "tikv,tidb";
```
4. query count(*)
```
 explain select count(*) from github_events_1;
ERROR 1105 (HY000): Can't find column test.github_events_1.test_count in schema Column: [Column#8] Unique key: []
```

### 2. What did you expect to see? (Required)

```
MySQL [test]> explain select count(*) from github_events_1;
+----------------------------+----------+-------------------+-----------------------+---------------------------------------------------------+
| id                         | estRows  | task              | access object         | operator info                                           |
+----------------------------+----------+-------------------+-----------------------+---------------------------------------------------------+
| StreamAgg_30               | 1.00     | root              |                       | funcs:count(Column#13)->Column#4                        |
| └─TableReader_31           | 1.00     | root              |                       | data:StreamAgg_10                                       |
|   └─StreamAgg_10           | 1.00     | cop[tikv]  |                       | funcs:count(test.github_events_1._tidb_rowid)->Column#13 |
|     └─TableFullScan_28     | 10000.00 | cop[tikv]  | table:github_events_1 | keep order:false, stats:pseudo                          |
+----------------------------+----------+-------------------+-----------------------+---------------------------------------------------------+
4 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
 explain select count(*) from github_events_1;
ERROR 1105 (HY000): Can't find column test.github_events_1.test_count in schema Column: [Column#8] Unique key: []
```

### 4. What is your TiDB version? (Required)

After tidb 6.4 , before tidb 6.5
master branch

### PR DESCRIPTION
**Pull request by elsa0520**
### What problem does this PR solve?

If the index and `count_star` optimizations exist at the same time, a `can't find column` error may occur.
The main reason is that the `count_star` optimization strategy selects a not null column, but due to the column is not added to the `colsRequiringFullLen` attribute, the planner mistakenly selects the index scan.
Eventually the previously selected column could not be found.

This PR fixed #39506 the problem above

Issue Number: close #39506

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

File: planner/core/rule_column_pruning.go [hunk 1]
```diff
@@ -348,8 +349,13 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *log
 		var handleCol *expression.Column
 		var handleColInfo *model.ColumnInfo
 		// case 1: tiflash
-		if ds.tableInfo.TiFlashReplica != nil {
+		_, isolationReadEnginesHasTiFlash := ds.ctx.GetSessionVars().GetIsolationReadEngines()[kv.TiFlash]
+		if ds.tableInfo.TiFlashReplica != nil && isolationReadEnginesHasTiFlash {
 			handleCol, handleColInfo = preferNotNullColumnFromTable(ds)
+			// The selected columns need to be added to `colsRequiringFullLen`.
```
Comment (AilinKid) [inline planner/core/rule_column_pruning.go:355]: the tiflash store shouldn't consider index? 
is `isolationReadEnginesHasTiFlash` enough?

Comment (elsa0520) [reply planner/core/rule_column_pruning.go:355]: I am afraid `isolationReadEnginesHasTiFlash ` is not enough.
It is possible that if both read engine and replica are satisfied, tikv is still used.

**Comment by elsa0520** [comment]
The master unit test is failed also.

**Review by fixdb — APPROVED** [review]
+1

File: planner/core/rule_column_pruning.go (hunk 1 above)
Comment (AilinKid) [reply planner/core/rule_column_pruning.go:355]: potential that we prefer a not null column while missing the index-single-scan candidate in TiKV storage read？

**Comment by elsa0520** [comment]
The count star has been changed from logical rule to post phase.
So the issue doesn't exist and this fix PR is no longer needed

**Pull request closed without merging**